xclim/cli.py

```diff
@@ -0,0 +1,239 @@
+# -*- coding: utf-8 -*-
+"""
+xclim command line interface module
+"""
+import inspect
+
+import click
+import xarray as xr
+from dask.diagnostics import ProgressBar
+
+import xclim as xc
+from xclim import atmos
+from xclim import land
+from xclim import seaIce
+
+
+xcmodules = {"atmos": atmos, "land": land, "seaIce": seaIce}
+
+
+def _get_indicator(indname):
+    if "." in indname:
+        modname, indname = indname.split(".")
+    else:
+        raise click.BadArgumentUsage(
+            f"Indicator name must include the module name (ex: atmos.tg_mean) (got {indname})"
+        )
+
+    if modname not in xcmodules or indname not in xcmodules[modname].__dict__:
+        raise click.BadArgumentUsage(
+            f"Indicator '{indname}' or module '{modname}' not found in xclim."
+        )
+
+    return xcmodules[modname].__dict__[indname]
+
+
+def _get_input(ctx):
+    arg = ctx.obj["input"]
+    if isinstance(arg, xr.Dataset):
+        return arg
+    elif isinstance(arg, tuple) or "*" in arg:
+        ctx.obj["xr_kwargs"].setdefault("combine", "by_coords")
+        ds = xr.open_mfdataset(arg, **ctx.obj["xr_kwargs"])
+    else:
+        ctx.obj["xr_kwargs"].pop("combine", None)
+        ds = xr.open_dataset(arg, **ctx.obj["xr_kwargs"])
+    ctx.obj["input"] = ds
+    return ds
+
+
+def _get_output(ctx):
+    if "ds_out" not in ctx.obj:
+        dsin = _get_input(ctx)
+        ctx.obj["ds_out"] = dsin.drop_vars(dsin.data_vars)
+    return ctx.obj["ds_out"]
+
+
+def _process_indicator(indicator, ctx, **params):
+    click.echo(
+        f"Processing : {indicator.format({'long_name': indicator.long_name}, params)['long_name']}"
+    )
+    dsin = _get_input(ctx)
+    dsout = _get_output(ctx)
+
+    for key, val in params.items():
+        if (
+            indicator._sig.parameters[key].default is inspect._empty
+        ):  # a Dataset is expected
+            if key == "tas" and val is None and key not in dsin.data_vars:
+                # Special case for tas.
+                params["tas"] = atmos.tg(dsin.tasmin, dsin.tasmax)
```

Comment:
Brittle. What guarantees that tasmin and tasmax are part of the inputs ? I would argue against special cases here.

Reply:
From what I've seen in the last months, it's quite common that `tas` is missing. I agree that this is brittle and I was going to add a check and warning before doing it, but I think it should be available.